README.md

# Documentation

[TOC]

## Overview

Aedis is a high-level [Redis](https://redis.io/) client library
built on top of
[Asio](https://www.boost.org/doc/libs/release/doc/html/boost_asio.html).
Some of its distinctive features are

* Support for the latest version of the Redis communication protocol [RESP3](https://github.com/redis/redis-specifications/blob/master/protocol/RESP3.md).
* Support for STL containers, TLS and Redis sentinel.
* Serialization and deserialization of your own data types.
* Healthy checks, back pressure and low latency.

In addition to that, Aedis hides most of the low-level Asio code away
from the user, which in the majority of the use cases will interact
with only three library entities

* `aedis::connection`: A healthy long-lasting connection to the Redis server.
* `aedis::resp3::request`: A container of Redis commands.
* `aedis::adapt()`: Adapts user data structures like STL containers to
  receive Redis responses.

Let us see how it works in more detail.

### Connection

The code below will establish a connection with a Redis
server where users can send commands (see intro.cpp)

```cpp
int main()
{
   net::io_context ioc;
   connection conn{ioc};

   conn.async_run({"127.0.0.1", "6379"}, {}, [](auto ec) { ... });

   // Pass conn to other operations ...

   ioc.run();
}
```

Requests on the other hand can be sent at any time, regardless of whether before or
after a connection was established. For example, the code below sends
the `PING` and `QUIT` commands, waits for the response and exits

```cpp
net::awaitable<void> ping(std::shared_ptr<connection> conn)
{
   // Request
   request req;
   req.push("PING", "some message");
   req.push("QUIT");

   // Response
   std::tuple<std::string, aedis::ignore> resp;

   // Execution
   co_await conn->async_exec(req, adapt(resp));
   std::cout << "Response: " << std::get<0>(resp) << std::endl;
}
```

The general structure about how to send commands is evident from the
code above

* Create a `aedis::resp3::request` object and add commands.
* Declare responses as elements of a `std::tuple`.
* Execute the request.

Multiple calls to `connection::async_exec` are synchronized
automatically so that different operations (or coroutines) don't have
to be aware of each other. Server side pushes can be received on the
same connection object that is being used to execute commands, for
example (see subscriber.cpp)

```cpp
net::awaitable<void> receive_pushes(std::shared_ptr<connection> conn)
{
   for (std::vector<node<std::string>> resp;;) {
      co_await conn->async_receive_push(adapt(resp));
      // Process the push in resp.
      resp.clear();
   }
}
```

@note Users should make sure any server pushes sent by the server are
consumed, otherwise the connection will eventually timeout.

### Reconnect

The `aedis::connection` class also supports reconnection.  In the
simplest scenario, after a connection lost users will want to
reconnect to the same server, the loop below shows how to do it

```cpp
net::awaitable<void> reconnect(std::shared_ptr<connection> conn)
{
   net::steady_timer timer{co_await net::this_coro::executor};
   for (;;) {
      boost::system::error_code ec;
      co_await conn->async_run({"127.0.0.1", "6379"}, {}, net::redirect_error(net::use_awaitable, ec));
      conn->reset_stream();
      timer.expires_after(std::chrono::seconds{1});
      co_await timer.async_wait();
   }
}
```
more complex scenarios, like performing a failover with sentinel can
be found in the examples. To aid proper failover, calls to
`connection::async_exec` won't automatically fail as a result of
connection lost, rather, they will remain suspended until a new
connection is established, once that happens all awaiting requests will be sent
automatically. This behaviour can be changed per request by setting on
the `aedis::resp3::request::config::close_on_connection_lost` or by
calling `connection::cancel()` with `connection::operation::exec`
which will cause all pending requests to be canceled.

### Timeouts

Aedis high-level API provides built-in support for most timeouts users
might need. For example, the `aedis::connection::async_run` member
function performs the following operations on behalf of the user

* Resolves Redis address.
* Connects to the endpoint.
* TLS handhshake (for TLS endpoints).
* RESP3 handshake, authentication and role check.
* Keeps sending PING commands to check for unresponsive servers.
* Keeps reading from the socket to handle server pushes and command responses.
* Keeps writing requests as it becomes possible e.g. after last response has arrived.

To control the timeout-behaviour of the operations above users must
create a `aedis::connection::timeouts` object and pass it to as
argument to the `aedis::connection::async_run` member function (or use
the suggested defaults).

Another related topic is the cancellation of
`aedis::connection::async_exec`.  With the introduction of awaitable
operators in Asio it is very simple implement timeouts either on
individual or on a group of operations, for example, users may be
tempted in writing code like

```cpp
co_await (conn.async_exec(...) || timer.async_wait(...))
```

The problem with this approach in Aedis is twofold

* Aedis has a buil-in healthy check that sends `PING` commands and
  checks whether responses are being received on time.  Since user
  commands use the same queue as the built-in `PING`, they are also
  subjected to the idle timeout, rendering cancellation like above
  unnecessary.

* To improve performance Redis encourages the use of pipelines, where
  many requests are sent in a single chunk to the server. In this
  scenario it is harder to cancel individual operations without
  causing all other (independent) requests in the same pipeline to
  fail too.

### Installation

Download the latest Aedis release from github 

```cpp
$ wget https://github.com/mzimbres/aedis/releases/download/v1.1.0/aedis-1.1.0.tar.gz
```

and unpack in your preferred location. Aedis is a header only
library, so you can starting using it. For that include the
following header 

```cpp
#include <aedis/src.hpp>

```
in no more than one source file in your applications (see intro.cpp
for example). To build the examples and run the tests cmake is also
supported

```cpp
$ BOOST_ROOT=/opt/boost_1_79_0/ cmake
$ make
$ make test
```

These are the requirements for using Aedis

- Boost 1.79 or greater.
- C++17. Some examples require C++20 with coroutine support.
- Redis 6 or higher. Optionally also redis-cli and Redis Sentinel.

The following compilers are supported

- Tested with gcc: 10, 11, 12.
- Tested with clang: 11, 13, 14.


### Examples

Users are encouraged to skim over the examples below before proceeding
to the next sections

* intro.cpp: The Aedis hello-world program. It sends one command to Redis and quits the connection.
* intro_tls.cpp: Same as intro.cpp but over TLS.
* intro_sync.cpp: Synchronous version of intro.cpp.
* containers.cpp: Shows how to send and receive stl containers and how to use transactions.
* serialization.cpp: Shows how to serialize types using Boost.Json.
* subscriber.cpp: Shows how to implement pubsub that reconnects and resubscribes when the connection is lost.
* subscriber_sentinel.cpp: Same as subscriber.cpp but with failover with sentinels.
* echo_server.cpp: A simple TCP echo server.
* chat_room.cpp: A simple chat room.

<a name="api-reference"></a>

### API Reference

* [High-Level](#high-level-api): Recommend to all users
* [Low-Level](#low-level-api): For users with needs yet to be imagined by the author.

In the next sections we will see how to create requests and receive
responses with more detail

## Requests

Redis requests are composed of one of more Redis commands (in
Redis documentation they are called
[pipelines](https://redis.io/topics/pipelining)). For example

```cpp
request req;

// Command with variable length of arguments.
req.push("SET", "key", "some value", value, "EX", "2");

// Pushes a list.
std::list<std::string> list
   {"channel1", "channel2", "channel3"};
req.push_range("SUBSCRIBE", list);

// Same as above but as an iterator range.
req.push_range2("SUBSCRIBE", std::cbegin(list), std::cend(list));

// Pushes a map.
std::map<std::string, mystruct> map
   { {"key1", "value1"}
   , {"key2", "value2"}
   , {"key3", "value3"}};
req.push_range("HSET", "key", map);
```

Sending a request to Redis is then performed with the following function

```cpp
co_await db->async_exec(req, adapt(resp));
```

<a name="serialization"></a>

### Serialization

The `push` and `push_range` functions above work with integers
e.g. `int` and `std::string` out of the box. To send your own
data type defined a `to_bulk` function like this

```cpp
struct mystruct {
   // Example struct.
};

void to_bulk(std::string& to, mystruct const& obj)
{
   std::string dummy = "Dummy serializaiton string.";
   aedis::resp3::to_bulk(to, dummy);
}
```

Once `to_bulk` is defined and accessible over ADL `mystruct` can
be passed to the `request`

```cpp
request req;

std::map<std::string, mystruct> map {...};

req.push_range("HSET", "key", map);
```

Example serialization.cpp shows how store json string in Redis.

<a name="responses"></a>

## Responses

To read responses effectively, users must know their RESP3 type,
this can be found in the Redis documentation for each command
(https://redis.io/commands). For example

Command  | RESP3 type                          | Documentation
---------|-------------------------------------|--------------
lpush    | Number                              | https://redis.io/commands/lpush
lrange   | Array                               | https://redis.io/commands/lrange

CHANGELOG.md
# Changelog

## master

* Removes `coalesce_requests` from the `connection::config`, this
  became a request property now, see `request::config::coalesce`.

* Removes `max_read_size` from the `connection::config`. The maximum
  read size can be specified now as a parameter of the
  `aedis::adapt()` function.

* Removes `aedis::sync` class, see intro_sync.cpp for how to perform
  synchronous and thread safe calls. This is possible in Boost. 1.80
  only as it requires `boost::asio::deferred`. 

* Moves from `boost::optional` to `std::optional`. This is part of
  moving to C++17.

* Changes the behaviour of the second `connection::async_run` overload
  so that it always return an error when the connection is lost.

* Adds TLS support, see intro_tls.cpp for how to use it.

* Adds an example that shows how to resolve addresses over sentinels,
  see subscriber_sentinel.cpp.

* Adds a `connection::config::resp3_handshake_timeout`. This is
  timeout used to send the `HELLO` command.

* Adds `endpoint` where in addition to host and port, users can
  optionally provide username, password and the expected server role
  (see `error::unexpected_server_role`).

* `connection::async_run` checks whether the server role received in
  the hello command is equal to the expected server role specified in
  `aedis::endpoint`. To skip this check let the role variable empty.

* Removes reconnect functionanlity from `aedis::connection`. It
  is possible in simple reconnection strategies but bloats the class
  in more complex scenarios, for example, with sentinel,
  authentication and TLS. This is trivial to implement in a separate
  coroutine. As a result the enum `event` and `async_receive_event`
  have been removed from the class too.

* Fixes a bug in `connection::async_receive_push` that prevented
  passing any response adapter other that `adapt(std::vector<node>)`.

* Changes the behaviour of `aedis::adapt()` that caused RESP3 errors
  to be ignored. One consequence of it is that `connection::async_run`
  would not exit with failure in servers that required authentication.

* Changes the behaviour of `connection::async_run` that would cause it
  to complete with success when an error in the
  `connection::async_exec` occurred.

* Ports the buildsystem from autotools to CMake.

## v1.0.0

* Adds experimental cmake support for windows users.

* Adds new class `aedis::sync` that wraps an `aedis::connection` in
  a thread-safe and synchronous API.  All free functions from the
  `sync.hpp` are now member functions of `aedis::sync`.

* Split `aedis::connection::async_receive_event` in two functions, one
  to receive events and another for server side pushes, see
  `aedis::connection::async_receive_push`.

* Removes collision between `aedis::adapter::adapt` and
  `aedis::adapt`.

* Adds `connection::operation` enum to replace `cancel_*` member
  functions with a single cancel function that gets the operations
  that should be cancelled as argument.

* Bugfix: a bug on reconnect from a state where the `connection` object
  had unsent commands. It could cause `async_exec` to never
  complete under certain conditions.

* Bugfix: Documentation of `adapt()` functions were missing from
  Doxygen.

## v0.3.0

* Adds `experimental::exec` and `receive_event` functions to offer a
  thread safe and synchronous way of executing requests across
  threads. See `intro_sync.cpp` and `subscriber_sync.cpp` for
  examples.

* `connection::async_read_push` was renamed to `async_receive_event`.

* `connection::async_receive_event` is now being used to communicate
  internal events to the user, such as resolve, connect, push etc. For
  examples see subscriber.cpp and `connection::event`.

* The `aedis` directory has been moved to `include` to look more
  similar to Boost libraries. Users should now replace `-I/aedis-path`
  with `-I/aedis-path/include` in the compiler flags.

* The `AUTH` and `HELLO` commands are now sent automatically. This change was
  necessary to implement reconnection. The username and password
  used in `AUTH` should be provided by the user on
  `connection::config`.

* Adds support for reconnection. See `connection::enable_reconnect`.

* Fixes a bug in the `connection::async_run(host, port)` overload
  that was causing crashes on reconnection.

* Fixes the executor usage in the connection class. Before theses
  changes it was imposing `any_io_executor` on users.

* `connection::async_receiver_event` is not cancelled anymore when
  `connection::async_run` exits. This change makes user code simpler.

* `connection::async_exec` with host and port overload has been
  removed. Use the other `connection::async_run` overload.

* The host and port parameters from `connection::async_run` have been
  move to `connection::config` to better support authentication and
  failover.


aedis/aedis.hpp
/* Copyright (c) 2018-2022 Marcelo Zimbres Silva (mzimbres@gmail.com)
 *
 * Distributed under the Boost Software License, Version 1.0. (See
 * accompanying file LICENSE.txt)
 */

#ifndef AEDIS_HPP
#define AEDIS_HPP

#include <aedis/error.hpp>
#include <aedis/adapt.hpp>
#include <aedis/connection.hpp>
#include <aedis/resp3/request.hpp>

// TODO: Support for Redis sentinel:
// 
//    - https://redis.io/docs/manual/sentinel and .
//    - https://redis.io/docs/reference/sentinel-clients
//
// TODO: Should we avoid conflicts of
//
//    - aedis::adapt and
//    - aedis::resp3::adapt
//
// function.

/** \mainpage Documentation
    \tableofcontents
  
    \section Overview
  
    Aedis is a high-level [Redis](https://redis.io/) client library
    built on top of [Asio](https://www.boost.org/doc/libs/release/doc/html/boost_asio.html)
    that provides simple and efficient communication with a Redis
    server. Some of its distinctive features are

    \li Support for the latest version of the Redis communication protocol [RESP3](https://github.com/redis/redis-specifications/blob/master/protocol/RESP3.md).
    \li First class support for STL containers and C++ built-in types.
    \li Serialization and deserialization of your own data types.
    \li Zero asymptotic allocations by means of memory reuse.
    \li Healthy checks, back pressure and low latency.

    Aedis API hides most of the low level asynchronous operations
    away from the user, for example, the code below pings a message to
    the server

    \code
    int main()
    {
       request req;
       req.push("HELLO", 3);
       req.push("PING");
       req.push("QUIT");

       std::tuple<aedis::ignore, std::string, aedis::ignore> resp;

       net::io_context ioc;

       connection db{ioc};

       db.async_exec("127.0.0.1", "6379", req, adapt(resp),
          [](auto ec, auto) { std::cout << ec.message() << std::endl; });

       ioc.run();

       // Print the ping message.
       std::cout << std::get<1>(resp) << std::endl;
    }
    \endcode

    For a detailed comparison of Redis clients and the design
    rationale behind Aedis jump to \ref why-aedis. For benchmarks see [](https://github.com/mzimbres/aedis/blob/master/benchmarks/benchmarks.md)

    \section requests Requests

    Redis requests are composed of one of more Redis commands (in
    Redis documentation they are called
    [pipelines](https://redis.io/topics/pipelining)). For example

    @code
    request req;

    // Command with variable length of arguments.
    req.push("SET", "key", "some value", value, "EX", "2");

    // Pushes a list.
    std::list<std::string> list
       {"channel1", "channel2", "channel3"};
    req.push_range("SUBSCRIBE", list);

    // Same as above but as an iterator range.
    req.push_range2("SUBSCRIBE", std::cbegin(list), std::cend(list));

    // Sends a map.
    std::map<std::string, mystruct> map
       { {"key1", "value1"}
       , {"key2", "value2"}
       , {"key3", "value3"}};
    req.push_range("HSET", "key", map);
    @endcode

    Sending a request to Redis is then peformed with the following function

    @code
    co_await db->async_exec(req, adapt(resp));
    @endcode

    The second argument \c adapt(resp) will be explained in \ref requests.

    \subsection requests-serialization Serialization

    The \c push and \c push_range functions above work with integers
    e.g. \c int and \c std::string out of the box. To send your own
    data type defined a \c to_bulk function like this

    @code
    struct mystruct {
       // Example struct.
    };
    
    void to_bulk(std::string& to, mystruct const& obj)
    {
       // Convert to obj string and call to_bulk.
       std::string dummy = "Dummy serializaiton string.";
       aedis::resp3::to_bulk(to, dummy);
    }
    @endcode

    Once \c to_bulk is defined and accessible over ADL \c mystruct can
    be passed to the \c request

    @code
    request req;

    std::map<std::string, mystruct> map {...};

    req.push_range("HSET", "key", map);
    @endcode

    It is quite common to store json string in Redis for example.

    \section low-level-responses Responses

    To read responses effectively, users must know their RESP3 type,
    this can be found in the Redis documentation for each command
    (https://redis.io/commands). For example

    Command  | RESP3 type                          | Documentation
    ---------|-------------------------------------|--------------
    lpush    | Number                              | https://redis.io/commands/lpush
    lrange   | Array                               | https://redis.io/commands/lrange
    set      | Simple-string, null or blob-string  | https://redis.io/commands/set
    get      | Blob-string                         | https://redis.io/commands/get
    smembers | Set                                 | https://redis.io/commands/smembers
    hgetall  | Map                                 | https://redis.io/commands/hgetall

    Once the RESP3 type of a given response is known we can choose a
    proper C++ data structure to receive it in. Fortunately, this is a
    simple task for most types. The table below summarises the options

    RESP3 type     | C++                                                          | Type
    ---------------|--------------------------------------------------------------|------------------
    Simple-string  | \c std::string                                             | Simple
    Simple-error   | \c std::string                                             | Simple
    Blob-string    | \c std::string, \c std::vector                             | Simple
    Blob-error     | \c std::string, \c std::vector                             | Simple
    Number         | `long long`, `int`, `std::size_t`, \c std::string          | Simple
    Double         | `double`, \c std::string                                   | Simple
    Null           | `boost::optional<T>`                                       | Simple
    Array          | \c std::vector, \c std::list, \c std::array, \c std::deque | Aggregate
    Map            | \c std::vector, \c std::map, \c std::unordered_map         | Aggregate
    Set            | \c std::vector, \c std::set, \c std::unordered_set         | Aggregate
    Push           | \c std::vector, \c std::map, \c std::unordered_map         | Aggregate

    Responses that contain nested aggregates or heterogeneous data
    types will be given special treatment later in \ref gen-case.  As
    of this writing, not all RESP3 types are used by the Redis server,
    which means in practice users will be concerned with a reduced
    subset of the RESP3 specification. Now let us see some examples

    @code
    // To ignore the response.
    co_await db->async_exec(req, adapt());

    // Read in a std::string e.g. get.
    std::string str;
    co_await db->async_exec(req, adapt(resp));

    // Read in a long long e.g. rpush.
    long long resp;
    co_await db->async_exec(req, adapt(resp));

    // Read in a std::set e.g. smembers.
    std::set<T, U> resp;
    co_await db->async_exec(req, adapt(resp));

    // Read in a std::map e.g. hgetall.
    std::map<T, U> resp;
    co_await db->async_exec(req, adapt(resp));

    // Read in a std::unordered_map e.g. hgetall.
    std::unordered_map<T, U> resp;
    co_await db->async_exec(req, adapt(resp));

    // Read in a std::vector e.g. lrange.
    std::vector<T> resp;
    co_await db->async_exec(req, adapt(resp));
    @endcode

    In other words, it is straightforward, just pass the result of \c
    adapt to the read function and make sure the response data type is
    compatible with the data structure you are calling @c adapter(...)
    with. All standard C++ containers are supported by Aedis.

    \subsection Optional

    It is not uncommon for apps to access keys that do not exist or
    that have already expired in the Redis server, to deal with these
    cases Aedis provides support for \c boost::optional. To use it,
    wrap your type around \c boost::optional like this

    @code
    boost::optional<std::unordered_map<T, U>> resp;
    co_await db->async_exec(req, adapt(resp));
    @endcode

    Everything else stays the same, before accessing data, users will
    have to check or assert the optional contains a value.

    \subsection heterogeneous_aggregates Heterogeneous aggregates

    There are cases where Redis returns aggregates that
    contain heterogeneous data, for example, an array that contains
    integers, strings nested sets etc. Aedis supports reading such
    aggregates in a \c std::tuple efficiently as long as the they
    don't contain 3-order nested aggregates e.g. an array that
    contains an array of arrays. For example, to read the response to
    a \c hello command we can use the following response type.

    @code
    using hello_type = std::tuple<
       std::string, std::string,
       std::string, std::string,
       std::string, int,
       std::string, int,
       std::string, std::string,
       std::string, std::string,
       std::string, std::vector<std::string>>;
    @endcode

    Transactions are another example where this feature is useful, for
    example, the response to the transaction below

    @code
    db.send("MULTI");
    db.send("GET", "key1");
    db.send("LRANGE", "key2", 0, -1);
    db.send("HGETALL", "key3");
    db.send("EXEC");
    @endcode

    can be read in the following way

    @code
    using trans_type = 
       std::tuple<
          boost::optional<std::string>, // get
          boost::optional<std::vector<std::string>>, // lrange
          boost::optional<std::map<std::string, std::string>> // hgetall
       >;

    std::tuple<
       aedis::ignore, // multi
       aedis::ignore, // get
       aedis::ignore, // lrange
       aedis::ignore, // hgetall
       trans_type, // exec
    > resp;

    co_await db->async_exec(req, adapt(resp));
    @endcode

    Note that above we are not ignoring the response to the commands
    themselves but whether they have been successfully queued.  Only
    after @c exec is received Redis will execute them in sequence and
    send all responses together in an array.

    \subsection Serialization

    As mentioned in \ref requests-serialization, it is common for
    users to serialized data before sending it to Redis e.g.  json
    strings, for example

    @code
    sr.push("SET", "key", "{"Server": "Redis"}"); // Unquoted for readability.
    sr.push("GET", "key")
    @endcode

    For performance and convenience reasons, we may want to avoid
    receiving the response to the \c get command above as a string
    just to convert it later to a e.g. deserialized json. To support
    this, Aedis calls a user defined \c from_bulk function while
    parsing the response. In simple terms, define your type

    @code
    struct mystruct {
       // struct fields.
    };
    @endcode

    and deserialize it from a string in a function \c from_bulk with
    the following signature

    @code
    void from_bulk(mystruct& obj, char const* p, std::size_t size, boost::system::error_code& ec)
    {
       // Deserializes p into obj.
    }
    @endcode

    After that, you can start receiving data efficiently in the desired
    types e.g. \c mystruct, \c std::map<std::string, mystruct> etc.

    \subsection gen-case The general case

    There are cases where responses to Redis
    commands won't fit in the model presented above, some examples are

    @li Commands (like \c set) whose response don't have a fixed
    RESP3 type. Expecting an \c int and receiving a blob-string
    will result in error.
    @li RESP3 aggregates that contain nested aggregates can't be read in STL containers.
    @li Transactions with a dynamic number of commands can't be read in a \c std::tuple.

    To deal with these cases Aedis provides the \c resp3::node
    type, that is the most general form of an element in a response,
    be it a simple RESP3 type or an aggregate. It is defined like this

    @code
    template <class String>
    struct node {
       // The RESP3 type of the data in this node.
       type data_type;

       // The number of elements of an aggregate (or 1 for simple data).
       std::size_t aggregate_size;

       // The depth of this node in the response tree.
       std::size_t depth;

       // The actual data. For aggregate types this is always empty.
       String value;
    };
    @endcode

    Any response to a Redis command can be received in a \c
    std::vector<node<std::string>>.  The vector can be seen as a
    pre-order view of the response tree
    (https://en.wikipedia.org/wiki/Tree_traversal#Pre-order,_NLR).
    Using it is no different that using other types

    @code
    // Receives any RESP3 simple data type.
    node<std::string> resp;
    co_await db->async_exec(req, adapt(resp));

    // Receives any RESP3 simple or aggregate data type.
    std::vector<node<std::string>> resp;
    co_await db->async_exec(req, adapt(resp));
    @endcode

    For example, suppose we want to retrieve a hash data structure
    from Redis with \c hgetall, some of the options are

    @li \c std::vector<node<std::string>: Works always.
    @li \c std::vector<std::string>: Efficient and flat, all elements as string.
    @li \c std::map<std::string, std::string>: Efficient if you need the data as a \c std::map
    @li \c std::map<U, V>: Efficient if you are storing serialized data. Avoids temporaries and requires \c from_bulk for \c U and \c V.

    In addition to the above users can also use unordered versions of the containers. The same reasoning also applies to sets e.g. \c smembers.

    \section examples Examples

    The examples listed below cover most use cases presented in the documentation above.

    @li intro.cpp: Basic steps with Aedis.
    @li containers.cpp: Shows how to send and receive stl containers.
    @li serialization.cpp: Shows the \c request support to serialization of user types.
    @li subscriber.cpp: Shows how channel subscription works.
    @li echo_server.cpp: A simple TCP echo server that users coroutines.
    @li chat_room.cpp: A simple chat room that uses coroutines.

    \section using-aedis Using Aedis

    To install and use Aedis you will need
  
    - Boost 1.78 or greater.
    - Unix Shell and Make (for linux users).
    - C++17. Some examples require C++20 with coroutine support.
    - Redis server.

    Some examples will also require interaction with
  
    - redis-cli: Used in one example.
    - Redis Sentinel Server: used in some examples.

    Aedis has been tested with the following compilers

    - Tested with gcc: 12, 11.
    - Tested with clang: 14, 13, 11.
  
    \section Installation

    The first thing to do is to download and unpack Aedis

    ```
    # Download the latest release on github
    $ wget https://github.com/mzimbres/aedis/releases
  
    # Uncompress the tarball and cd into the dir
    $ tar -xzvf aedis-version.tar.gz
    ```

    If you can't use \c configure and \c make (e.g. Windows users)
    add the directory where you unpacked Aedis to the
    include directories in your project, otherwise run
  
    ```
    # See configure --help for all options.
    $ ./configure --prefix=/opt/aedis-version --with-boost=/opt/boost_1_78_0

    # Install Aedis in the path specified in --prefix
    $ sudo make install
  
    ```
  
    and include the following header 
  
    ```cpp
    #include <aedis/src.hpp>

    ```

    in exactly one source file in your applications. At this point you
    can start using Aedis. To build the examples and run the tests run
  
    ```
    # Build aedis examples.
    $ make examples
  
    # Test aedis in your machine.
    $ make check
    ```

    \section Developers
  
    To generate the build system run
  
    ```
    $ autoreconf -i
    ```
  
    After that you will have a configure script 
    that you can run as explained above, for example, to use a
    compiler other that the system compiler run
  
    ```
    $ CC=/opt/gcc-10.2.0/bin/gcc-10.2.0 CXX=/opt/gcc-10.2.0/bin/g++-10.2.0 CXXFLAGS="-g -Wall -Werror"  ./configure ...
    $ make distcheck
    ```

    \section why-aedis Why Aedis

    At the time of this writing there are seventeen Redis clients
    listed in the [official](https://redis.io/docs/clients/#cpp) list.
    With so many clients available it is not unlikely that users are
    asking themselves why yet another one.  In this section I will try
    to compare Aedis to the most popular clients and why we need
    Aedis. Notice however that this is ongoing work as comparing
    client objectively is difficult and time consuming.

    The most popular client at the moment of this writing ranked by
    github stars is

    @li https://github.com/sewenew/redis-plus-plus

    Before we start it is worth mentioning some of the things it does
    not support

    @li RESP3. Without RESP3 is impossible to support some important Redis features like client side caching, among other things.
    @li The Asio asynchronous model.
    @li Reading response diretly in user data structures avoiding temporaries.
    @li Error handling with error-code and exception overloads.
    @li Healthy checks.

    The remaining points will be addressed individually.

    @subsection redis-plus-plus

    Let us first have a look at what sending a command a pipeline and a
    transaction look like

    @code
    auto redis = Redis("tcp://127.0.0.1:6379");

    // Send commands
    redis.set("key", "val");
    auto val = redis.get("key"); // val is of type OptionalString.
    if (val)
        std::cout << *val << std::endl;

    // Sending pipelines
    auto pipe = redis.pipeline();
    auto pipe_replies = pipe.set("key", "value")
                            .get("key")
                            .rename("key", "new-key")
                            .rpush("list", {"a", "b", "c"})
                            .lrange("list", 0, -1)
                            .exec();

    // Parse reply with reply type and index.
    auto set_cmd_result = pipe_replies.get<bool>(0);
    // ...

    // Sending a transaction
    auto tx = redis.transaction();
    auto tx_replies = tx.incr("num0")
                        .incr("num1")
                        .mget({"num0", "num1"})
                        .exec();

    auto incr_result0 = tx_replies.get<long long>(0);
    // ...
    @endcode

    Some of the problems with this API are

    @li Heterogeneous treatment of commands, pipelines and transaction.
    @li Any Api that sends individual commands has a very restricted scope of usability and should be avoided for performance reasons.
    @li The API imposes exceptions on users, no error-code overload is provided.
    @li No way to reuse the buffer for new calls to e.g. \c redis.get in order to avoid further dynamic memory allocations.
    @li Error handling of resolve and connection no clear.

    According to the documentation, pipelines in redis-plus-plus have
    the following characteristics

    > NOTE: By default, creating a Pipeline object is NOT cheap, since
    > it creates a new connection.

    This is clearly a downside of the API as pipelines should be the
    default way of communicating and not an exception, paying such a
    high price for each pipeline imposes a severe cost in performance.
    Transactions also suffer from the very same problem

    > NOTE: Creating a Transaction object is NOT cheap, since it
    > creates a new connection.

    In Aedis there is no difference between sending one command, a
    pipeline or a transaction because creating the request is decoupled
    from the IO objects.

    > redis-plus-plus also supports async interface, however, async
    > support for Transaction and Subscriber is still on the way.
    > 
    > The async interface depends on third-party event library, and so
    > far, only libuv is supported.

    Async code in redis-plus-plus looks like the following

    @code
    auto async_redis = AsyncRedis(opts, pool_opts);
    
    Future<string> ping_res = async_redis.ping();
    
    cout << ping_res.get() << endl;
    @endcode

    As the reader can see, the async interface is based on futures
    which is also known to have a bad performance.  The biggest
    problem however with this async design is that it makes it
    impossible to write asynchronous programs correctly since it
    starts an async operation on every command sent instead of
    enqueueing a message and triggering a write when it can be sent.
    It is also not clear how are pipelines realised with the design
    (if at all).

    \section Acknowledgement

    Some people that were helpful in the development of Aedis

    @li Richard Hodges ([madmongo1](https://github.com/madmongo1)): For helping me with Asio and the design of asynchronous programs in general.
    @li Vinícius dos Santos Oliveira ([vinipsmaker](https://github.com/vinipsmaker)): For useful discussion about how Aedis consumes buffers in the read operation (among other things).
    @li Petr Dannhofer ([Eddie-cz](https://github.com/Eddie-cz)): For helping me understand how the `AUTH` and `HELLO` command can influence each other.

    \section Reference
  
    See \subpage any.

 */

/** \defgroup any Reference
 *
 *  This page contains the documentation of all user facing code.
 */

#endif // AEDIS_HPP
